### PR TITLE
sound/soc/aml/m8: work-in-progress

### PR DESCRIPTION
**Pull request by Kwiboo**
This is work-in-progress for improving audio on S905:

- No need to truncate `addr` and `size`, dma buffer should already be 4096 aligned
- No need to reset `AIU_MEM_I2S/IEC958_BUF_CNTL`, we are not using buffer level
- Period and buffer bytes is now more aligned to hw fifo size
- Use default copy method now that buffer is better aligned
- Period timer is now reset when buffer pointers is reset
- SPDIF 32-bit is now bit shifted same amount as I2S
- I2S 24-bit should no longer be bit shifted

This should make stereo LPCM 16/24/32-bit work on both I2S and SPDIF dai and multi channel LPCM 24/32-bit work on I2S dai (16 bit needs more work), note that 32-bit audio is truncated to 24-bit by hw

Remaining work:
- Remove the now unneeded second buffer
- Set channel status values from AES params

@LongChair please test this with mpv
@lrusak I think this should make hifi shield and hifi shield plus work using the I2S device

### CODE REVIEW

**Comment by lrusak** [comment]
Thanks @Kwiboo. Can you look at my changes here https://github.com/LibreELEC/linux-amlogic/compare/amlogic-3.14.y...lrusak:hifi-hat I'm wondering about my changes to aml-i2s-dai.c and how the sound card actually gets activated. If you are able to pick these changes and fix them up that would be great as my knowledge in this area is limited.

**Comment by Kwiboo** [comment]
The HK changes mostly adds two new cards, switches to use interleaved write/read mode (we already do this with "split-mode") and enables higher sample rates.

There is no special activation as far as I know, the i2s driver is already sending data to the i2s interface, the only thing done in aml-i2s-dai.c is making it possible to disable hdmi out when the "special" dac-card is used. Our current code should already send audio data to both i2s and hdmi but lack any option to disable hdmi.

My guess is that hifi shield and hifi shield plus or any other i2s card already "works" when device 0 is uses, but should work a little bit better with this for 16/24-bit audio.

To fully take advantage of hifi shield 2 we need to import some HK code as it can be controlled with i2c/spi.

**Comment by kszaq** [comment]
On a S905X device which uses T9015 codec for I2S audio I'm getting the following error:

> [    3.710733@0] aml_codec_T9015 c8832000.t9015: aml_T9015_audio_codec_probe
> [    3.717582@0] aml_pcm_dai: pcm mode detection =1
> [    3.749465@0] input input1: key 116 up
> [    3.929670@0] aml_snd_card: headphone detection disable=1
> [    3.929806@0] aml_snd_m8_card aml_m8_snd.46: alloc playback DMA buffer error
> [    3.936429@0] aml-i2s i2s_platform.42: ASoC: pcm constructor failed: -12
> [    3.943088@0] aml_snd_m8_card aml_m8_snd.46: ASoC: can't create pcm I2S.39 :-12
> [    3.950353@0] aml_snd_m8_card aml_m8_snd.46: ASoC: failed to instantiate card -12
> [    3.957971@0] aml_snd_m8_card aml_m8_snd.46: register aml sound card error -12
> [    3.964940@0] aml_snd_m8_card aml_m8_snd.46: Can't probe snd_soc_card
> [    3.971303@0] aml_snd_m8_card: probe of aml_m8_snd.46 failed with error -12

**Comment by lrusak** [comment]
any updates?

**Comment by barturblits** [comment]
Hi there,
I am wondering if you guys are also working on hard(er) limiting the volume gain from +24 to 0 dB?
At the moment it is set in the alsamixer to 205 or 207 and the user changes volume through pulse in ubuntu for odroid c2. But i would like to disable pulse and use alsa directly. This makes it dangerous to change volume. I saw the guys from hifiberry limited the volume with snd_soc_limit_volume for pcm512x  so you have the alsamixer slider at 0dB at 100% and through a switch you can explicitly enable +24dB gain again.

**Comment by fat-tony** [comment]
Guys,
I'm using kszaq's builds for the S905 (1GB RAM / 8GB ROM / 1Gb ethernet) and am having some issues with high bit rate stereo and multichannel FLAC audio playback. I have my DVD-Audio and SACD collection ripped to FLAC (24bit). Most multichannel is at 96kHz or 88.2kHz and plays back fine over HDMI. I have several stereo rips at 192kHz or 176.4kHz but they don't produce any audio stream to my Sony AV amp though the Kodi logs show 2 channel 192 (or 176.4)kHz being selected. If I choose a fixed output sample rate of 192kHz Kodi plays back a multichannel output according to my AV amp (left and right channels correct although the centre and rear channels etc. are silent). It seems somehow that a two channel selection by Kodi is somehow invalid but 6 or 8 channels are ok. Additionally, I note that I can't choose a fixed sample rate output on 176.4kHz in the Kodi dropdown - only 44.1, 88.2, 96 and 192kHz. Is there a known issue with high bit rate stereo output on these Amlogic devices? I was using an S905X initially, but kszaq advised that there was a 96kHz upper limit on audio sampling rate and that the S905 was better for audio, so I switched. Any pointers gratefully appreciated! I can supply logs or test audio if necessary.

**Comment by Ray-future** [comment]
@fat-tony: This is not the place to discuss your issues.

**Comment by fat-tony** [comment]
Apologies if I've overstepped the mark! High bit rate multichannel audio has a limited user base so I though the devs might be interested in real-world usage issues. I helped out with testing during development of linux audio engine for Kodi a couple of years back. No audio issues for me on LibreELEC running on my Asus ION board, so I thought maybe the Amlogic platform might be equally capable. Where can I raise this issue, please?

**Comment by Ray-future** [comment]
And we appreciate your help. LibreELEC Forum is used for Bugreports iirc.

File: sound/soc/aml/m8/aml_i2s.c [hunk 1]
```diff
@@ -90,14 +90,14 @@ static const struct snd_pcm_hardware aml_i2s_hardware = {
...
-	.period_bytes_min = 64,
-	.period_bytes_max = 32 * 1024 * 2,
-	.periods_min = 2,
-	.periods_max = 1024,
-	.buffer_bytes_max = 128 * 1024 * 2 * 2,
+	.period_bytes_min = 256,
+	.period_bytes_max = 32 * 1024,
+	.periods_min = 4,
+	.periods_max = 64,
+	.buffer_bytes_max = 32 * 1024 * 64,
```
Comment (kszaq) [inline sound/soc/aml/m8/aml_i2s.c:97]: Please consider reverting this line as it introduces issues S905X using T9015 codec.

**Comment by fat-tony** [comment]
@Kwiboo - many thanks for recent patches used by @kszaq. Has resolved all high bit-rate stereo audio (176.4 and 192kHz) for me with @kszaq's builds for S905/S905X. I appreciate your work!

**Comment by chewitt** [comment]
Is this work still valid, or has it been (or will it be) superseded by other things? - update please :)

**Comment by Kwiboo** [comment]
I think @kszaq and/or @Raybuntu may have used parts of this in their builds, they may know if this is still valid or should be superseded, I have lost interest in ancient Amlogic kernels.

**Comment by Ray-future** [comment]
I'm ok with merging this.

**Comment by kszaq** [comment]
Some issues were reported when I had this commit in my community builds, mainly with audio passthrough. If interest is lost, I suggest closing this PR, current audio code is fine for regular use.

**Comment by Ray-future** [comment]
Closing this as it has conflicts and @Kwiboo is no longer interested working on this.

**Pull request closed without merging**